server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,263 @@
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public long fileLength(String name) throws IOException {
+        logger.trace("fileLength() called {}", name);
+        readLock.lock();
+        try {
+            long fileLength;
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
+                fileLength = localDirectory.fileLength(name);
+                logger.trace("fileLength from Local {}", fileLength);
+            } else {
+                fileLength = remoteDirectory.fileLength(name);
+                logger.trace("fileLength from Remote {}", fileLength);
+            }
+            return fileLength;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        logger.trace("createOutput() called {}", name);
+        writeLock.lock();
+        try {
+            return localDirectory.createOutput(name, context);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public IndexOutput createTempOutput(String prefix, String suffix, IOContext context) throws IOException {
+        logger.trace("createTempOutput() called {} , {}", prefix, suffix);
+        writeLock.lock();
+        try {
+            return localDirectory.createTempOutput(prefix, suffix, context);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        writeLock.lock();
+        try {
+            Collection<String> newLocalFiles = new ArrayList<>();
+            Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+            for (String name : names) {
+                if (remoteFiles.contains(name) == false) newLocalFiles.add(name);
+            }
+            logger.trace("Synced files : {}", newLocalFiles);
+            localDirectory.sync(newLocalFiles);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void syncMetaData() throws IOException {
+        logger.trace("syncMetaData() called ");
+        writeLock.lock();
+        try {
+            localDirectory.syncMetaData();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        writeLock.lock();
+        try {
+            localDirectory.rename(source, dest);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        writeLock.lock();
+        try {
+            if (Arrays.asList(getRemoteFiles()).contains(name) == false) {
```

Comment:
Instead of checking `(Arrays.asList(getRemoteFiles()).contains(name) == false)`, shouldn't we check if the file is present locally?

Reply:
where do we check if the file is present in the cache?